flink-runtime/src/main/java/org/apache/flink/runtime/state/filesystem/FsCheckpointStorageAccess.java

```diff
@@ -232,7 +232,9 @@ public FsMergingCheckpointStorageAccess toFileMergingStorage(
             FileMergingSnapshotManager mergingSnapshotManager, Environment environment)
             throws IOException {
         return new FsMergingCheckpointStorageAccess(
-                fileSystem,
+                /* Multiple subtask/threads would share one output stream,
+                SafetyNetWrapperFileSystem cannot be used to prevent different threads from interfering with each other when exiting. */
+                FileSystem.getUnguardedFileSystem(checkpointsDirectory.toUri()),
```

Comment:
Also, I have a concern here: When we use `GuardedFileSystem`, all streams can be closed properly when the corresponding thread is exiting. After this change, if we use `UnguardedFileSystem`, it may cause stream leak.

Reply:
> I'm thinking could checkpointsDirectory as the parameter of FsMergingCheckpointStorageAccess constructor?

👍 Agreed, I will update the constructor as your suggestion.

> Also, I have a concern here: When we use GuardedFileSystem, all streams can be closed properly when the corresponding thread is exiting. After this change, if we use UnguardedFileSystem, it may cause stream leak.

`FileMergingSnapshotManager`  takes this responsibility, `FileMergingSnapshotManager` would close all held physical files and streams when the task executor close/shutdown.

Reply:
> FileMergingSnapshotManager takes this responsibility, FileMergingSnapshotManager would close all held physical files and streams when the task executor close/shutdown.

Sorry, I'm not familiar with `FileMergingSnapshotManager`, thanks for the clarification. Sounds good to me.